vm/vm.go

```diff
@@ -2,12 +2,102 @@
 package vm
 
 import (
+	"fmt"
+	"sync"
+
+	peer "github.com/libp2p/go-libp2p-peer"
 	"github.com/libp2p/go-smart-record/ir"
 )
 
+// RecordValue determines the structure of data stored in a record.
+// Each peer has a private dataspace to store dicts.
+type RecordValue map[peer.ID]*ir.Dict
+
+type Record struct {
+	Key   string      // Key of the record
+	Value RecordValue // Root stores peers dictionaries in that key.
+}
+
 // Machine captures the public interface of a smart record virtual machine.
 type Machine interface {
-	Update(ir.Dict) error
-	Query(ir.Dict) error
-	Get() ir.Dict
+	Update(writer peer.ID, k string, update ir.Dict) error // Updates the dictionary in the writer's private space.
+	Get(k string) Record                                   // Get the full Record in a key
+	// NOTE: No query operation will be supported until we figure out selectors
+	// Query(key string, selector Selector) (RecordValue, error)
+}
+
+// VM implements the Machine interface and keeps the map of records in its state.
+type vm struct {
+	ctx ir.UpdateContext // UpdateContext the VM uses to resolve conflicts
+	//ds  ds.Datastore    // TODO: Add a datastore instead of using map[string] for the VM state
+	keys map[string]*Record // State of the VM storing the map of records.
+	asm  ir.Assembler       // Assemble to use in the VM.
+	lk   sync.RWMutex       // Lock to enable multiple access
+}
+
+// NewVM creates a new smart record Machine
+func NewVM(ctx ir.UpdateContext, asm ir.Assembler) Machine {
+	return newVM(ctx, asm)
+}
+
+//newVM instantiates a new VM with an updateContext and an assembler
+func newVM(ctx ir.UpdateContext, asm ir.Assembler) *vm {
+	return &vm{
+		ctx:  ctx,
+		keys: make(map[string]*Record),
+		asm:  asm,
+	}
+}
+
+// Get the whole record stored in a key
+func (v *vm) Get(k string) Record {
+	v.lk.RLock()
+	defer v.lk.RUnlock()
+	if v.keys[k] == nil {
+		return Record{Key: k, Value: RecordValue{}}
```

Comment:
The VM communicates (via Update and Get) with the "outside world" (the caller) only via syntactic nodes (not semantic nodes). The vm structure here keeps semantic values, because it stores them after they have been assembled (in Update below). 

Therefore, before Get returns a value it must convert it into a syntactic one first.

This points out an oversight on my part: I need to move the Disassemble() method from Smart to Node (and actually entirely get rid of Smart interface). I'll do this and ping you.

Reply:
OK. Now all nodes can disassemble: https://github.com/libp2p/go-smart-record/commit/da19a960dce6673b60a12eb4ecf5a0167ba96fd6

Reply:
This was actually a question I had. This solves it. Thanks!